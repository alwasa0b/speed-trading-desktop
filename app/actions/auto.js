import {
  UPDATE_TIME_INTERVAL,
  UPDATE_UNDER_BID_PRICE,
  UPDATE_OVER_ASK_PRICE,
  UPDATE_QUANTITY_AUTO,
  NUMBER_OF_RUNS,
  UPDATE_NUMBER_OF_OPEN_ORDERS,
  UPDATE_PAUSE_PRICE,
  CLEAR
} from "../constants/auto";

import {
  START_WORKER,
  WORKER_STARTED,
  STOP_WORKER,
  WORKER_STOPPED,
  PAUSE_WORKER,
  UPDATE_WORKER,
  WORKER_UPDATED,
  WORKER_PAUSED,
  WORKER_RESUMED,
  RESUME_WORKER,
  PANIC
} from "../constants/messages";

const { ipcRenderer } = require("electron");

export const update_pause_price = payload => ({
  type: UPDATE_PAUSE_PRICE,
  payload
});

export const update_number_of_runs = payload => ({
  type: NUMBER_OF_RUNS,
  payload
});

export const update_number_of_open_orders = payload => ({
  type: UPDATE_NUMBER_OF_OPEN_ORDERS,
  payload
});

export const update_time_interval = payload => ({
  type: UPDATE_TIME_INTERVAL,
  payload
});

export const update_under_bid_price = payload => ({
  type: UPDATE_UNDER_BID_PRICE,
  payload
});

export const update_over_ask_price = payload => ({
  type: UPDATE_OVER_ASK_PRICE,
  payload
});

export const update_quantity = payload => ({
  type: UPDATE_QUANTITY_AUTO,
  payload
});

export const startStopWorker = () => (dispatch, getState) => {
  const { auto_order } = getState();
  if (auto_order.running) stop_worker(dispatch, getState);
  else start_worker(dispatch, getState);
};

const start_worker = async (dispatch, getState, ipc = ipcRenderer) => {
  const { auto_order, messages } = getState();
  const { instrument, symbol, price, high } = messages.price;

  ipc.once(WORKER_STARTED, async event => {
    dispatch({ type: WORKER_STARTED });
  });

  ipc.once(WORKER_STOPPED, async event => {
    dispatch({ type: WORKER_STOPPED });
  });

  await ipc.send(START_WORKER, {
    ...auto_order,
    messages: [],
    instrument,
    symbol,
    price,
    high
  });
};

const stop_worker = async (dispatch, getState, ipc = ipcRenderer) => {
  const { auto_order, messages } = getState();
  const { instrument, symbol } = messages.price;

  await ipc.send(STOP_WORKER, {
    ...auto_order,
    instrument,
    symbol
  });
};

export const pause_resume_worker = () => (dispatch, getState) => {
  const { auto_order } = getState();
  if (auto_order.paused) resume_worker(dispatch, getState);
  else pause_worker(dispatch, getState);
};

const resume_worker = async (dispatch, getState, ipc = ipcRenderer) => {
  await ipc.send(RESUME_WORKER);
};

const pause_worker = async (dispatch, getState, ipc = ipcRenderer) => {
  ipc.once(WORKER_PAUSED, async event => {
    dispatch({ type: WORKER_PAUSED });
  });
  ipc.once(WORKER_RESUMED, async event => {
    dispatch({ type: WORKER_RESUMED });
  });
  await ipc.send(PAUSE_WORKER);
};

export const update_worker = () => async (
  dispatch,
  getState,
  ipc = ipcRenderer
) => {
  const { auto_order, messages } = getState();
  const { instrument, symbol } = messages.price;

  ipc.once(WORKER_UPDATED, async event => {
    dispatch({ type: WORKER_UPDATED });
  });

  await ipc.send(UPDATE_WORKER, {
    ...auto_order,
    instrument,
    symbol
  });
};

export const panic = () => async (dispatch, getState, ipc = ipcRenderer) => {
  await ipc.send(PANIC);
};

export const clear = () => ({ type: CLEAR });
