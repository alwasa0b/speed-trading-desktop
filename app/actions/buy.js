import {
  PLACE_BUY_REQUEST_SUCCESS,
  UPDATE_QUANTITY,
  UPDATE_SELL_ORDER_TYPE,
  UPDATE_BUY_ORDER_TYPE,
  UPDATE_BUY_PRICE,
  UPDATE_SELL_PRICE,
  PLACE_BUY_REQUEST
} from "../constants/buy";
const { ipcRenderer } = require("electron");

const buy_order_success = () => ({ type: PLACE_BUY_REQUEST_SUCCESS });
export const place_buy_order = () => async (dispatch, getState) => {
  const { buy_order, messages } = getState();
  const { instrument, symbol } = messages.price;

  await ipcRenderer.send(PLACE_BUY_REQUEST, {
    ...buy_order,
    instrument,
    symbol
  });

  ipcRenderer.once(PLACE_BUY_REQUEST_SUCCESS, async (event, data) => {
    dispatch(buy_order_success(data));
  });
};

export const update_quantity = ({ quantity }) => ({
  type: UPDATE_QUANTITY,
  quantity
});

export const update_sell_order_type = ({ sell_order_type }) => ({
  type: UPDATE_SELL_ORDER_TYPE,
  sell_order_type
});

export const update_buy_order_type = ({ buy_order_type }) => ({
  type: UPDATE_BUY_ORDER_TYPE,
  buy_order_type
});

export const update_buy_price = ({ buy_price }) => ({
  type: UPDATE_BUY_PRICE,
  buy_price
});

export const update_sell_price = ({ sell_price }) => ({
  type: UPDATE_SELL_PRICE,
  sell_price
});
