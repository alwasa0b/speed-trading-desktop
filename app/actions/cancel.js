import {
  PLACE_CANCEL_REQUEST,
  PLACE_CANCEL_REQUEST_SUCCESS
} from "../constants/cancel";

import { ipcRenderer } from "electron";

const cancel_order_success = () => ({ type: PLACE_CANCEL_REQUEST_SUCCESS });
export const place_cancel_order = ({ cancel_order }) => async dispatch => {
  await ipcRenderer.send(PLACE_CANCEL_REQUEST, cancel_order);

  ipcRenderer.once(PLACE_CANCEL_REQUEST_SUCCESS, async (event, data) => {
    dispatch(cancel_order_success(data));
  });
};
