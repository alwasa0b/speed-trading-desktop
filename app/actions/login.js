import {
  LOGIN_REQUEST,
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  UPDATE_USERNAME,
  UPDATE_PASSWORD
} from "../constants/login";

import { ipcRenderer } from "electron";

const login_success = () => ({ type: LOGIN_SUCCESS });
const login_failed = () => ({ type: LOGIN_FAILURE });

export const login = () => async (dispatch, getState) => {
  await ipcRenderer.send(LOGIN_REQUEST, { login: getState().login });
  ipcRenderer.once(LOGIN_SUCCESS, async (event, data) => {
    dispatch(login_success());
  });
};

export const update_username = ({ username }) => ({
  type: UPDATE_USERNAME,
  username
});

export const update_password = ({ password }) => ({
  type: UPDATE_PASSWORD,
  password
});
