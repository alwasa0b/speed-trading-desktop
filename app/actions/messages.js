import {
  UPDATE_POSITIONS,
  UPDATE_ORDERS,
  UPDATE_PRICE,
  PRICE_UPDATED,
  POSITIONS_UPDATED,
  ORDERS_UPDATED
} from "../constants/messages";

const { ipcRenderer } = require("electron");

export const update_positions = () => async (
  dispatch,
  getState,
  ipc = ipcRenderer
) => {
  await ipc.removeAllListeners(POSITIONS_UPDATED);
  await ipc.send(UPDATE_POSITIONS);

  ipc.on(POSITIONS_UPDATED, async (event, data) => {
    dispatch({ type: POSITIONS_UPDATED, data });
  });
};

export const update_orders = () => async (
  dispatch,
  getState,
  ipc = ipcRenderer
) => {
  await ipc.removeAllListeners(ORDERS_UPDATED);
  await ipc.send(UPDATE_ORDERS);

  ipc.on(ORDERS_UPDATED, async (event, data) => {
    dispatch({ type: ORDERS_UPDATED, data });
  });
};

export const update_price = ({ symbol }) => async (
  dispatch,
  getState,
  ipc = ipcRenderer
) => {
  await ipc.removeAllListeners(PRICE_UPDATED);
  await ipc.send(UPDATE_PRICE, { symbol });

  ipc.on(PRICE_UPDATED, async (event, data) => {
    dispatch({ type: PRICE_UPDATED, data });
  });
};
