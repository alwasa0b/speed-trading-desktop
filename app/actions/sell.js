import {
  PLACE_SELL_REQUEST_SUCCESS,
  PLACE_SELL_REQUEST,
  UPDATE_SELL_ORDER_TYPE_TYPE,
  UPDATE_SELL_ORDER_PRICE_PRICE
} from "../constants/sell";
import { ipcRenderer } from "electron";

const sell_order_success = () => ({ type: PLACE_SELL_REQUEST_SUCCESS });
export const place_sell_order = ({
  sell_order,
  price,
  order_type
}) => async dispatch => {
  const { instrument, quantity, symbol } = sell_order;

  await ipcRenderer.send(PLACE_SELL_REQUEST, {
    price,
    order_type,
    quantity,
    instrument,
    symbol
  });

  ipcRenderer.once(PLACE_SELL_REQUEST_SUCCESS, async (event, data) => {
    dispatch(sell_order_success(data));
  });
};

export const update_sell_order_type = ({ order_type, symbol }) => {
  return { type: UPDATE_SELL_ORDER_TYPE_TYPE, order_type, symbol };
};

export const update_sell_order_price = ({ price, symbol }) => {
  return { type: UPDATE_SELL_ORDER_PRICE_PRICE, price, symbol };
};
