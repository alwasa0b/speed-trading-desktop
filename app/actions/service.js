import axios from "axios";

const API_ROOT =
  process.env.NODE_ENV === "production"
    ? `${window.location.origin}`
    : "http://localhost:3001";

const post = async (endpoint, data) => {
  try {
    // if (endpoint === "login" && sessionStorage.getItem("sp"))
    //   return { response: {}, data };

    const response = await axios.post(`${API_ROOT}/${endpoint}`, data);
    
    // if (endpoint === "login") {
    //   // sessionStorage.setItem("sp", response.payload.data.token);
    // }

    return { response, data };
  } catch (e) {
    return { detail: e.toString() };
  }
};

export const login = async ({ login }) => await post("login", login);
export const place_cancel_order = async ({ cancel_order }) => await post("place_cancel_order", cancel_order);
export const place_sell_order = async ({ sell_order }) => await post("place_sell_order", sell_order);
export const place_buy_order = async ({ buy_order }) => await post("place_buy_order", buy_order);
export const place_stop_loss_order = async ({ stop_order }) => await post("place_stop_loss_order", stop_order);