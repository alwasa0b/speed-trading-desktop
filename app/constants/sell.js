export const PLACE_SELL_REQUEST = "PLACE_SELL_REQUEST";
export const PLACE_SELL_REQUEST_SUCCESS = "PLACE_SELL_REQUEST_SUCCESS";
export const UPDATE_SELL_ORDER_TYPE_TYPE = "UPDATE_SELL_ORDER_TYPE_TYPE";
export const UPDATE_SELL_ORDER_PRICE_PRICE = "UPDATE_SELL_ORDER_PRICE_PRICE";
