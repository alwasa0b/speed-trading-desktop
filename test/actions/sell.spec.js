import sinon, { spy } from "sinon";
import * as actions from "../../app/actions/sell";
import {
  PLACE_SELL_REQUEST_SUCCESS,
  PLACE_SELL_REQUEST
} from "../../app/constants/sell";

describe("sell actions", () => {
  it("update_sell_order_type should create update_sell_order_type action", () => {
    expect(
      actions.update_sell_order_type({
        order_type: "test_type",
        symbol: "test_symbol"
      })
    ).toMatchSnapshot();
  });

  it("update_sell_order_price should create update_sell_order_price action", () => {
    expect(
      actions.update_sell_order_price({
        price: 100,
        symbol: "test_symbol"
      })
    ).toMatchSnapshot();
  });

  it("update_quantity should create update_quantity action", () => {
    expect(
      actions.update_quantity({
        quantity: 100,
        symbol: "test_symbol"
      })
    ).toMatchSnapshot();
  });

  it("update_quantity_type should create update_quantity_type action", () => {
    expect(
      actions.update_quantity_type({
        quantity_type: "test_quantity_type",
        symbol: "test_symbol",
        quantity: 100
      })
    ).toMatchSnapshot();
  });

  it("should call ipc.send with sell_order", () => {
    const fn = actions.place_sell_order({
      sell_order: {
        instrument: "instrument",
        symbol: "symbol",
        quantity: "total_quantity",
        shares_held_for_sells: "shares_held_for_sells"
      },
      price: 0,
      order_type: "order_type",
      quantity: "requested_quantity",
      quantity_type: "order_type"
    });
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    const ipc = { send: spy(), once: spy() };

    fn(dispatch, () => {}, ipc);

    expect(
      ipc.send.calledWith(PLACE_SELL_REQUEST, {
        price: 0,
        order_type: "order_type",
        requested_quantity: "requested_quantity",
        instrument: "instrument",
        symbol: "symbol",
        quantity_type: "order_type",
        total_quantity: "total_quantity",
        shares_held_for_sells: "shares_held_for_sells"
      })
    ).toBe(true);
  });

  it("should start listening to ipc.once", done => {
    const fn = actions.place_sell_order({
      sell_order: {
        instrument: "instrument",
        symbol: "symbol",
        quantity: "total_quantity",
        shares_held_for_sells: "shares_held_for_sells"
      },
      price: 0,
      order_type: "order_type",
      quantity: "requested_quantity",
      quantity_type: "order_type"
    });
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    const ipc = { send: spy(), once: spy() };

    fn(dispatch, () => {}, ipc);

    setTimeout(() => {
      expect(
        ipc.once.calledWith(
          PLACE_SELL_REQUEST_SUCCESS,
          sinon.match.typeOf("function")
        )
      ).toBe(true);
      done();
    }, 5);
  });
});
