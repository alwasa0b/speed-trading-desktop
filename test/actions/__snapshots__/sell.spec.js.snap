// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sell actions update_quantity should create update_quantity action 1`] = `
Object {
  "quantity": 100,
  "symbol": "test_symbol",
  "type": "UPDATE_SELL_ORDER_QUANTITY",
}
`;

exports[`sell actions update_quantity_type should create update_quantity_type action 1`] = `
Object {
  "quantity": 100,
  "quantity_type": "test_quantity_type",
  "symbol": "test_symbol",
  "type": "UPDATE_SELL_ORDER_QUANTITY_TYPE",
}
`;

exports[`sell actions update_sell_order_price should create update_sell_order_price action 1`] = `
Object {
  "price": 100,
  "symbol": "test_symbol",
  "type": "UPDATE_SELL_ORDER_PRICE_PRICE",
}
`;

exports[`sell actions update_sell_order_type should create update_sell_order_type action 1`] = `
Object {
  "order_type": "test_type",
  "symbol": "test_symbol",
  "type": "UPDATE_SELL_ORDER_TYPE_TYPE",
}
`;
